MovingNonOpfsFiles.md

```diff
@@ -0,0 +1,224 @@
+
+## What about moving files from the OPFS to user visible directories, or vice-versa?
+
+Previous proposals included support for best-effort moves of files and directories across file systems. We are not pursuing this at this time.
+
+### What challenges exist with moving files out of the OPFS?
+
+* Files in the OPFS [have few restrictions on the allowed characters in their names](https://github.com/web-platform-tests/wpt/blob/4981b40a9b00f87091c417e096e40c327b9407ed/fs/script-tests/FileSystemDirectoryHandle-getFileHandle.js#L18-L44), while [local files are limited by what’s allowed on the underlying operating system](https://fs.spec.whatwg.org/#valid-file-name).
+* Files written within the OPFS are not subject to security checks, while [user agents are encouraged](https://wicg.github.io/file-system-access/#security-malware) to perform safe browsing checks and apply the Mark-of-the-Web to local files created or modified by this API. Note that this challenge is significantly more daunting for directory moves.
+
+These challenges are all resolvable, but not without a compelling use case for OPFS &lt;-> local file moves.
+
+### What about exporting files from the OPFS to the local file system?
+
+A common use case cited in earlier proposals was to “export” a file from the OPFS to the local file system. In practice, we expect most sites just want to _copy_ the file to the local file system and retain a copy in the OPFS.
+
+Besides, since files within the OPFS may or may not correspond to “actual” files on disk, this move operation is likely to be a create + copy + delete anyways and would likely come with marginal performance gains, if any.
+
+For now, we don’t believe this is a use case which requires built-in API support. Developers can create + copy + delete (as they do today - see the code snippet above).
+
+### What if I have a compelling use case?
+
+If a compelling use case comes along, we can always reconsider this decision and add support later. It’s much easier to add functionality to the web platform than to remove it. To make your case, please file an issue on the spec at <https://github.com/WICG/file-system-access/issues>.
+
+### What about moving files between OPFS instances?
+
+The [Storage Buckets API](https://wicg.github.io/storage-buckets/explainer.html) will allow a site to have multiple Origin Private File System instances (whoops, [that name aged poorly](http://go/gh/whatwg/fs/issues/92)). Since that feature is still in incubation, we are not considering this at this time.
+
+### What if I try it anyways?
+
+The promise will be rejected with a `NotSupportedError``DOMException`.
```

Comment:
https://github.com/whatwg/fs/issues/57#issuecomment-1277022240 suggests `NotSupportedError` is "for not-yet-launched features", so perhaps this should be `InvalidModificationError` if we never expect to support these moves (["the object cannot be modified in this way"](https://developer.mozilla.org/en-US/docs/Web/API/DOMException))